lsp/src/Language/LSP/Server/Processing.hs

```diff
@@ -449,31 +452,35 @@ handle' ::
   TClientMessage meth ->
   m ()
 handle' logger mAction m msg = do
-  maybe (return ()) (\f -> f msg) mAction
+  shutdown <- isShuttingDown
+  -- These are the methods that we are allowed to process during shutdown.
+  -- The reason that we include 'shutdown' itself here is because we change 
+  -- state to "shutting down" just _before_ we process the message itself!
+  let allowedMethod m = case (splitClientMethod m, m) of
+        (IsClientNot, SMethod_Exit) -> True
+        -- we do not allow SMethod_Shutdown to be called during shutdown
```

Comment:
I think we should mark SMethod_Shutdown as false, so it won't be called twice.

Reply:
That won't work, because we toggle `isShuttingDown` before we process the handler! So then we won't run the user's shutdown handler...

Reply:
Yes, done